### DISCUSSION TITLE
Help with webhook for remote updates

### DISCUSSION THREAD

**Discussion by datwaft** [abraunegg/onedrive#1844] [Q&A] [answered]
Hello!

I am currently trying to configure OneDrive to support remote updates via webhook.

I have succesfully configured OneDrive until now to work with remote mode and as a system service, so I know that the problem is related to the webhook.

I first tried without using the nginx proxy, and the following error is displayed:

```console
foo@bar:~$ onedrive --monitor
Configuration file successfully loaded
config file has been updated, checking if --resync needed
Configuring Global Azure AD Endpoints
Initializing the Synchronization Engine ...
Initializing monitor ...
OneDrive monitor interval (seconds): 300
Initializing subscription for updates ...
onedrive.OneDriveException@src/onedrive.d(1629): HTTP request returned status code 400 (Bad Request)
{
    "error": {
        "code": "InvalidRequest",
        "innerError": {
            "client-request-id": "5e71d956-99e5-41a6-a0ac-079581a87a21",
            "date": "2022-02-16T20:31:24",
            "request-id": "5e71d956-99e5-41a6-a0ac-079581a87a21"
        },
        "message": "NotificationUrl scheme='http' is not supported."
    }
}
...
```

The same error happened after setting up the nginx reverse proxy with the following configuration:

```conf
server {
  listen 80;
  location /webhooks/onedrive {
    proxy_pass http://127.0.0.1:8888;
  }
}
```

After that I thought about trying with self-signed SSL certificate.

Here is the nginx configuration:

```conf
server {
  listen 443 ssl;

  ssl_certificate /etc/nginx/ssl/reverse-proxy.crt;
  ssl_certificate_key /etc/nginx/ssl/reverse-proxy.key;

  location /webhooks/onedrive {
    proxy_pass http://127.0.0.1:8888;
  }
}
```

Here is the error:

```console
foo@bar:~$ onedrive --monitor
Configuration file successfully loaded
config file has been updated, checking if --resync needed
Configuring Global Azure AD Endpoints
Initializing the Synchronization Engine ...
Initializing monitor ...
OneDrive monitor interval (seconds): 300
Initializing subscription for updates ...
onedrive.OneDriveException@src/onedrive.d(1629): HTTP request returned status code 400 (Bad Request)
{
    "error": {
        "code": "InvalidRequest",
        "innerError": {
            "client-request-id": "3e34e6f8-35f3-42e0-a25a-bbdb1a51f44b",
            "date": "2022-02-16T20:36:29",
            "request-id": "3e34e6f8-35f3-42e0-a25a-bbdb1a51f44b"
        },
        "message": "The underlying connection was closed: Could not establish trust relationship for the SSL\/TLS secure channel."
    }
}
```

Do you have any idea what could be the problem? I couldn't get certificate signed by https://letsencrypt.org/ as I don't have a domain.

**Comment by abraunegg** [comment, marked as answer]
@datwaft 

What client version are you trying this with? The 400 error response should be correctly handled - but it appears it is not. Please open a bug report for this following the correct support process for this particular issue.

> Do you have any idea what could be the problem? I couldn't get certificate signed by https://letsencrypt.org/ as I don't have a domain.

You need to have a valid domain + valid certificate - without this, it is not going to work because if you are using any self-signed certificates etc - how is Microsoft going to 'trust' that - they cant ..

**Reply by datwaft** [reply]
> You need to have a valid domain + valid certificate - without this, it is not going to work because if you are using any self-signed certificates etc - how is Microsoft going to 'trust' that - they cant ..

Indeed, I suspected that was the issue. Thanks!

> What client version are you trying this with? The 400 error response should be correctly handled - but it appears it is not. Please open a bug report for this following the correct support process for this particular issue.

I am using v2.4.15-16-g3c119d3. It should be latest version as I just installed a few days ago from source.

Later today I will open the issue.

**Reply by datwaft** [reply]
Btw, I got a domain using https://freenom.com/ and it seems to work properly with the certificate.

**Comment by abraunegg** [comment]
@datwaft 
> Btw, I got a domain using https://freenom.com/ and it seems to work properly with the certificate.

Thanks for the feedback.

If you can 'revert' your changes - if possible - and test the following PR - this should also trap the 400 error message:

```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/1846/head:pr1846
git checkout pr1846
./configure; make clean; make;
```
To run the PR, you need to run the client from the PR build directory:
```
./onedrive <any options needed>
```

When running the PR, your version should be: `onedrive v2.4.15-19-gf642463` or greater.

**Reply by datwaft** [reply]
It seems to work properly:
![image](https://user-images.githubusercontent.com/37723586/154367995-eb4e97aa-b092-4741-9c59-e1f71ef062b8.png)

Here is the log copied from the terminal:
```console
$ ./onedrive --version
onedrive v2.4.15-19-gf642463
$ ./onedrive --monitor                                         
Configuration file successfully loaded
Configuring Global Azure AD Endpoints
Initializing the Synchronization Engine ...
Initializing monitor ...
OneDrive monitor interval (seconds): 300
Initializing subscription for updates ...

ERROR: Microsoft OneDrive API returned an error with the following message:
  Error Message:    HTTP request returned status code 400 (Bad Request)
  Error Reason:     NotificationUrl scheme='http' is not supported.
  Error Timestamp:  2022-02-16T22:26:42
  API Request ID:   c35af7b9-ca5a-40de-b5f9-5e27666ec4d9
  Calling Function: createSubscription()

std.json.JSONException@std/json.d(308): JSONValue is not an object
----------------
/home/datwaft/dlang/dmd-2.098.1/linux/bin64/../../src/phobos/std/exception.d:516 pure @safe noreturn std.exception.bailOut!(std.json.JSONException).bailOut(immutable(char)[], ulong, scope const(char)[]) [0x55bab31852bf]
/home/datwaft/dlang/dmd-2.098.1/linux/bin64/../../src/phobos/std/exception.d:437 pure @safe bool std.exception.enforce!(std.json.JSONException).enforce!(bool).enforce(bool, lazy const(char)[], immutable(char)[], ulong) [0x55bab3185233]
??:? inout pure @property @trusted inout(std.json.JSONValue[immutable(char)[]]) std.json.JSONValue.objectNoRef() [0x55bab32398c9]
??:? inout pure ref @safe inout(std.json.JSONValue) std.json.JSONValue.opIndex(return immutable(char)[]) [0x55bab3239acc]
src/onedrive.d:1016 void onedrive.OneDriveApi.createSubscription() [0x55bab31bf68b]
src/onedrive.d:979 void onedrive.OneDriveApi.createOrRenewSubscription() [0x55bab31bf2db]
src/main.d:1173 _Dmain [0x55bab31b3dbb]
std.net.curl.CurlException@std/net/curl.d(4402): A libcurl function was given a bad argument on handle null
----------------
??:? pure @safe noreturn std.exception.bailOut!(std.net.curl.CurlException).bailOut(immutable(char)[], ulong, scope const(char)[]) [0x55bab3294ca6]
??:? pure @safe bool std.exception.enforce!(std.net.curl.CurlException).enforce!(bool).enforce(bool, lazy const(char)[], immutable(char)[], ulong) [0x55bab3294c13]
??:? void std.net.curl.Curl._check(int) [0x55bab3260065]
??:? void std.net.curl.Curl.clear(etc.c.curl.CurlOption) [0x55bab326041a]
??:? void std.net.curl.HTTP.clearRequestHeaders() [0x55bab325f1ab]
src/onedrive.d:490 void onedrive.OneDriveApi.shutdown() [0x55bab31bc4e1]
src/main.d:64 _Dmain [0x55bab31b4850]
```

**Reply by abraunegg** [reply]
Thanks .. OK .. one sec .. will update the PR

**Comment by abraunegg** [comment]
@datwaft 
Please can you rebuild the PR using the same instructions - the client version should now be `onedrive v2.4.15-20-g9a346c3` and that crash handled.

**Reply by datwaft** [reply]
Here is the output:

![image](https://user-images.githubusercontent.com/37723586/154369616-147ba548-e80b-4cf1-88f1-36aacaddd0fe.png)

```console
$ ./onedrive --version
onedrive v2.4.15-20-g9a346c3
$ ./onedrive --monitor
Configuration file successfully loaded
Configuring Global Azure AD Endpoints
Initializing the Synchronization Engine ...
Initializing monitor ...
OneDrive monitor interval (seconds): 300
Initializing subscription for updates ...

ERROR: Microsoft OneDrive API returned an error with the following message:
  Error Message:    HTTP request returned status code 400 (Bad Request)
  Error Reason:     NotificationUrl scheme='http' is not supported.
  Error Timestamp:  2022-02-16T22:40:02
  API Request ID:   25945dd0-548a-475b-81f7-b9be5a1cae1b
  Calling Function: createSubscription()

ERROR: Unable to initialize subscriptions for updates. Please fix this issue.
```

**Comment by datwaft** [comment]
Btw, I think it would be helpful to add to the USAGE.md that you need to have SSL certificate, as the example uses `http` and `http` is not supported.

> With nginx running, you can configure `webhook_public_url` to `http://<your_host>/webhooks/onedrive`.

Additionally, I tried to use the snippet provided in that file and it didn't work with my nginx version. I had to remove the http part:

```diff
- http {
    server {
      listen 80;
      location /webhooks/onedrive {
        proxy_pass http://127.0.0.1:8888;
      }
    }
- }
```

**Comment by datwaft** [comment]
Question: should I still open the issue or just with the PR is okay?

**Reply by abraunegg** [reply]
@datwaft 
> Question: should I still open the issue or just with the PR is okay?

No - should be OK - as the issue is fixed, just need to update the doco as per your other finding, then merge that into 'master'
